lib/FunctionIsolation/IsolatedFunctionKind.cpp

```diff
@@ -0,0 +1,184 @@
+#include "revng/Yield/CrossRelations.h"
+#include "revng/Yield/Pipes/ProcessCallGraph.h"
+
+using namespace pipeline;
+using namespace ::revng::kinds;
+using namespace llvm;
+using yield::CrossRelations;
+
+static bool
+haveFSOAndPreservedRegsChanged(const model::TypePath &OldPrototype,
+                               const model::TypePath &NewPrototype) {
+  auto Old = abi::FunctionType::Layout::make(OldPrototype);
+  auto New = abi::FunctionType::Layout::make(NewPrototype);
+
+  return Old.FinalStackOffset != New.FinalStackOffset
+         or Old.CalleeSavedRegisters != New.CalleeSavedRegisters;
+}
+
+using IsolatedFK = IsolatedFunctionKind;
+void IsolatedFK::getInvalidations(const Context &Ctx,
+                                  TargetsList &ToRemove,
+                                  const GlobalTupleTreeDiff &Diff,
+                                  const pipeline::Global &Before,
+                                  const pipeline::Global &After) const {
+  using PM = PathMatcher;
+  using Binary = model::Binary;
+  using FunctionAttribute = model::FunctionAttribute::Values;
+
+  if (not llvm::isa<TupleTreeGlobal<Binary>>(Before))
+    return;
+
+  std::set<MetaAddress> Targets;
+  const auto &OldModel = cast<TupleTreeGlobal<Binary>>(Before).get();
+  const auto &Model = cast<TupleTreeGlobal<Binary>>(After).get();
+
+  // CrossRelations container not found? Take the most conservative approach,
+  // namely, to invalidate all targets.
+  if (not Ctx.containsReadOnlyContainer(BinaryCrossRelationsRole)) {
+    invalidateAllTargetsPerFunctionRank(Model, Targets);
+    return;
+  }
+
+  static constexpr auto &BCRR = BinaryCrossRelationsRole;
+  using CRFileContainer = pipes::CrossRelationsFileContainer;
+  const auto &Container = Ctx.getReadOnlyContainer<CRFileContainer>(BCRR);
+  if (Container.empty()) {
+    invalidateAllTargetsPerFunctionRank(Model, Targets);
+    return;
+  }
+
+  const TupleTree<CrossRelations> &CrossRelations = Container.get();
+  const TupleTreeDiff<Binary> *ModelDiff = Diff.getAs<Binary>();
+  if (not ModelDiff)
+    return;
+
+  auto FuncAttrMatcher = PM::create<Binary>("/Functions/*/Attributes").value();
+  auto DynFuncAttrMatcher = PM::create<Binary>("/ImportedDynamicFunctions/*/"
+                                               "Attributes")
+                              .value();
+  const auto &IDF = Model->ImportedDynamicFunctions;
+
+  for (const Change<Binary> &Change : ModelDiff->Changes) {
+    revng_assert(not Change.Path.empty());
+
+    // A new `model::Function` has been added or removed?
+    if (pathAsString<Binary>(Change.Path) == "/Functions") {
+      if (Change.New) {
+        // Invalidate all targets
+        invalidateAllTargetsPerFunctionRank(Model, Targets);
+      } else if (Change.Old) {
+        // Invalidate function's callers
+        auto Function = std::get<model::Function>(*Change.Old);
+        insertCallers(CrossRelations, Function.Entry, Targets);
```

Comment:
Why not `insertCallersAndTransitiveClosureIfInline`?

Reply:
As we have said, when removing a function, querying the `Relations` fields of `BinaryCrossRelations` should suffice in order to retrieve the callers, there should be no need to traverse the graph.

Reply:
If a caller is inline, we need to recompute its callers too.